src/Evolution/DiscontinuousGalerkin/SlopeLimiters/HwenoModifiedSolution.cpp

```diff
@@ -0,0 +1,342 @@
+    for (const auto& dir_and_neighbors : element.neighbors()) {
+      result.push_back(dir_and_neighbors.first);
+    }
+    return result;
+  }
+  ();
+
+  for (const auto& dir : directions_with_neighbors) {
+    interpolation_matrices[dir] =
+        volume_interpolation_matrix(mesh, mesh, element, dir);
+    quadrature_weights_dot_interpolation_matrices[dir] =
+        vector_dot_matrix(quadrature_weights, interpolation_matrices.at(dir));
+  }
+
+  for (const auto& primary_dir : directions_with_neighbors) {
+    if (directions_with_neighbors.size() == 1) {
+      // With a single neighbor, there can be no neighbors to exclude.
+      std::vector<Direction<VolumeDim>> nothing_to_exclude{{}};
+      // To reuse the same data structure from the more common `else` branch,
+      // here we stick the data into the (normally nonsensical) slot where
+      // `dir_to_exclude == primary_dir`.
+      a_inverse_matrices[primary_dir][primary_dir] = a_inverse_matrix(
+          element, mesh, quadrature_weights, interpolation_matrices,
+          quadrature_weights_dot_interpolation_matrices, primary_dir,
+          nothing_to_exclude);
+    } else {
+      // With multiple neighbors, there can be 1+ neighbors to exclude. Because
+      // the case where 2+ secondary neighbors must be excluded is rare, we do
+      // not cache it. So we only handle the case of 1 neighbor to exclude.
+      for (const auto& dir_to_exclude : directions_with_neighbors) {
+        // Skip the nonsensical case where the primary and excluded neighbors
+        // are the same.
+        if (dir_to_exclude == primary_dir) {
+          continue;
+        }
+        a_inverse_matrices[primary_dir][dir_to_exclude] = a_inverse_matrix(
+            element, mesh, quadrature_weights, interpolation_matrices,
+            quadrature_weights_dot_interpolation_matrices, primary_dir,
+            {{dir_to_exclude}});
+      }
+    }
+  }
+}
+
+template <size_t VolumeDim>
+const Matrix& ConstrainedFitCache<VolumeDim>::retrieve_a_inverse_matrix(
+    const Direction<VolumeDim>& primary_direction,
+    const std::vector<Direction<VolumeDim>>& directions_to_exclude) const
+    noexcept {
+  if (LIKELY(directions_to_exclude.size() == 1)) {
+    // The typical case
+    return a_inverse_matrices.at(primary_direction)
+        .at(directions_to_exclude[0]);
+  } else if (directions_to_exclude.empty()) {
+    // The atypical case is cached in a normally nonsensical slot to reuse the
+    // data structure.
+    return a_inverse_matrices.at(primary_direction).at(primary_direction);
+  } else {
+    ERROR("Asking for an uncached matrix, with multiple excluded neighbors");
+  }
+}
+
+namespace {
+
+template <size_t VolumeDim, size_t DummyIndex>
+const ConstrainedFitCache<VolumeDim>& constrained_fit_cache_impl(
+    const Element<VolumeDim>& element, const Mesh<VolumeDim>& mesh) noexcept {
+  static const ConstrainedFitCache<VolumeDim> result(element, mesh);
```

Comment:
This (in general, picked a fairly arbitrary line to comment on) appears to assume that the mesh is the same in every element.  I don't think this is true for all our domains, although it probably is for all of the physically interesting ones.  I think this is a reasonable thing to rely upon until p-refinement is supported, but if it is easy to put a sanity check for it somewhere that would be good to have.

Reply:
Agreed; I've added another check.